flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -110,7 +116,12 @@ protected void performDefaultAction(DefaultActionContext context) throws Excepti
 		final LegacySourceFunctionThread sourceThread = new LegacySourceFunctionThread(getName());
 		sourceThread.start();
 		sourceThread.getCompletionFuture().whenComplete((Void ignore, Throwable sourceThreadThrowable) -> {
-			if (sourceThreadThrowable == null) {
+			if (sourceThreadThrowable == null || isFinished) {
+
+				// the isFinished is only set in the case of SYNC_SAVEPOINT. In this case, the final savepoint is
+				// already completed, so we do not risk the POISON_PILL to overpass the savepoint barrier and lead
+				// to a deadlock.
```

Comment:
Personally, I'd remove this comment, as it may be irrelevant or become outdated.
The comment near the `isFinished` field already has some explanation wrt exceptions.